### PR TITLE
TMOP PA diagonal kernels with non-diagonal targets.

### PR DESCRIPTION
**Pull request by vladotomov**
The `Gradient::AssembleDiagonal` PA kernels in TMOP were assuming that the target matrices `W` were diagonal. This is not the case for `GIVEN_SHAPE_AND_SIZE` targets given by a mesh that has non-square elements. The `W` propagates in the formulas as it multiplies the shape gradients from `T = A Winv`.

This PR fixes the diagonal 2D/3D kernels and adds unit tests that show the issue.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4015](https://github.com/mfem/mfem/pull/4015) | @vladotomov | @tzanio | @kmittal2 + @v-dobrev + @camierjs | 12/7/23 | 3/28/24 | 4/2/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
@kmittal2, @camierjs and @v-dobrev -- please review

**Review by kmittal2 — APPROVED** [review]

File: tests/unit/miniapps/test_tmop_pa.cpp
Comment (kmittal2) [inline tests/unit/miniapps/test_tmop_pa.cpp:1]: Not related to this PR: We don't really use the `lin_solver=0` or `1` option (in the unit test or in practice generally) so we could consider removing them.

**Comment by tzanio** [comment]
@mstees and @camierjs -- can you please try to review this week?

**Review by mstees — APPROVED** [review]

**Review by camierjs — APPROVED** [review]

**Comment by v-dobrev** [comment]
I'm trying to test with CUDA and it turns out the new kernels use too much shared memory, so compilation fails for the 3D kernel.

In 3D, the kernels in `master` (which are not generally valid) use just $2$ shared memory (SM) doubles per quadrature point, so they are able to fit the default max quad points of $14^3$ (for CUDA) -- total is $(2\times 14^3) \times 8 = 43904$ bytes. The max on V100 is $49152$ bytes of SM.

The current (generally valid) implementation in 3D uses $3^3+2\times 3^2 = 45$ SM doubles per quadrature point, so the max quad points it can fit is $5^3$. This is using $45\times 5^3\times 8 = 45000$ bytes.

I think we can improve things and reduce the storage per quadrature point to $2\times 3^2 = 18$ SM doubles. With that we'll be able to fit $6^3$ quad points, using $18\times 6^3\times 8 = 31104$ bytes. We can squeeze in $7^3$ quad points if we limit the polynomial degree to $\le 5$. In this case, we can use one array of size $3^2\times 7^3$ and one array of size $3^2\times 7^2\times 6$ for a total of $45864$ bytes.

**Comment by v-dobrev** [comment]
I think I resolved the issue with the CUDA/ROCM build -- someone else should take another look at these changes.

**Comment by v-dobrev** [comment]
One remark:

When I run the following: (these tests are not run by `make test` -- it runs these executables without `--all`)
```console
./tmop_pa_tests_cpu --all --enable-output --verbose
./tmop_pa_tests_debug --all --enable-output --verbose
```
I see failures, however, the issue does not seem related to this PR -- I see the same issue in `master`.

**Review by v-dobrev — APPROVED** [review]

**Pull request merged**